File: inc.graphql-resources/user-profile.php
<?php

require_once __DIR__ . '/database.php';
require_once __DIR__ . '/auth.php';
require_once __DIR__ . '/inc/I18n.php';

// --- end i18n ---

// Handle POST request for updating user profile
if ($_SERVER['REQUEST_METHOD'] === 'POST') {
    header('Content-Type: application/json');
    require_once __DIR__ . '/database.php';

    try {
        $sql = "UPDATE admin SET name = :name, email = :email, gender = :gender, birth_day = :birth_day, phone = :phone WHERE username = :username";
        $stmt = $db->prepare($sql);
        $stmt->execute(array(
            ':name' => htmlspecialchars(isset($_POST['name']) ? $_POST['name'] : ''),
            ':email' => htmlspecialchars(isset($_POST['email']) ? $_POST['email'] : ''),
            ':gender' => htmlspecialchars(isset($_POST['gender']) ? $_POST['gender'] : ''),
            ':birth_day' => htmlspecialchars(isset($_POST['birth_day']) ? $_POST['birth_day'] : ''),
            ':phone' => htmlspecialchars(isset($_POST['phone']) ? $_POST['phone'] : ''),
            ':username' => $_SESSION['username'] // Use session username for security
        ));

        echo json_encode(array('success' => true, 'message' => $i18n->t('profile_updated_successfully')));
    } catch (Exception $e) {
        http_response_code(500);
        echo json_encode(array('success' => false, 'message' => $i18n->t('failed_to_update_profile', $e->getMessage())));
    }
    exit;
}



/*
CREATE TABLE IF NOT EXISTS admin (
	 admin_id NVARCHAR(40) PRIMARY KEY NOT NULL,
	 name NVARCHAR(100) NULL,
	 username NVARCHAR(100) NULL,
	 password NVARCHAR(512) NULL,
	 password_version NVARCHAR(512) NULL,
	 admin_level_id NVARCHAR(40) NULL,
	 gender NVARCHAR(2) NULL,
	 birth_day DATE NULL,
	 email NVARCHAR(100) NULL,
	 phone NVARCHAR(100) NULL,
	 language_id NVARCHAR(40) NULL,
	 validation_code TEXT NULL,
	 last_reset_password TIMESTAMP NULL,
	 blocked BOOLEAN NULL DEFAULT 0,
	 time_create TIMESTAMP NULL,
	 time_edit TIMESTAMP NULL,
	 admin_create NVARCHAR(40) NULL,
	 admin_edit NVARCHAR(40) NULL,
	 ip_create NVARCHAR(50) NULL,
	 ip_edit NVARCHAR(50) NULL,
	 active BOOLEAN NULL DEFAULT 1
);
*/
try
{
$sql = "SELECT admin.*, (SELECT admin_level.name FROM admin_level WHERE admin_level.admin_level_id = admin.admin_level_id) AS admin_level_name FROM admin WHERE admin.username = :username";
$stmt = $db->prepare($sql);
$stmt->execute(array(':username' => $_SESSION['username']));
$admin = $stmt->fetch(PDO::FETCH_ASSOC);

if (isset($_GET) && isset($_GET['action']) && $_GET['action'] == 'update') {
?>
    <div class="table-container detail-view">
        <form id="profile-update-form" class="form-group" onsubmit="handleProfileUpdate(event); return false;">
            <table class="table table-borderless">
                <tr>
                    <td><?php echo $i18n->t('admin_id'); ?></td>
                    <td><input type="text" name="admin_id" class="form-control" value="<?php echo htmlspecialchars($admin['admin_id']); ?>" readonly></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('name'); ?></td>
                    <td><input type="text" name="name" class="form-control" value="<?php echo htmlspecialchars($admin['name']); ?>"></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('username'); ?></td>
                    <td><input type="text" name="username" class="form-control" value="<?php echo htmlspecialchars($admin['username']); ?>" readonly></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('gender'); ?></td>
                    <td>
                        <select name="gender" class="form-control">
                            <option value="M" <?php echo ($admin['gender'] == 'M') ? 'selected' : ''; ?>><?php echo $i18n->t('male'); ?></option>
                            <option value="F" <?php echo ($admin['gender'] == 'F') ? 'selected' : ''; ?>><?php echo $i18n->t('female'); ?></option>
                        </select>
                    </td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('birthday'); ?></td>
                    <td><input type="date" name="birth_day" class="form-control" value="<?php echo htmlspecialchars($admin['birth_day']); ?>"></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('phone'); ?></td>
                    <td><input type="text" name="phone" class="form-control" value="<?php echo htmlspecialchars($admin['phone']); ?>"></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('email'); ?></td>
                    <td><input type="email" name="email" class="form-control" value="<?php echo htmlspecialchars($admin['email']); ?>"></td>
                </tr>
                <tr>
                    <td></td>
                    <td>
                        <button type="submit" class="btn btn-success"><?php echo $i18n->t('update'); ?></button>
                        <button type="button" class="btn btn-secondary" onclick="window.location='#user-profile'"><?php echo $i18n->t('cancel'); ?></button>
                    </td>
                </tr>
            </table>
        </form>
    </div>

<?php
} else {
?>
    <div class="table-container detail-view">
        <form action="" class="form-group">
            <table class="table table-borderless">
                <tr>
                    <td><?php echo $i18n->t('admin_id'); ?></td>
                    <td><?php echo htmlspecialchars($admin['admin_id']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('name'); ?></td>
                    <td><?php echo htmlspecialchars($admin['name']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('username'); ?></td>
                    <td><?php echo htmlspecialchars($admin['username']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('gender'); ?></td>
                    <td><?php echo ($admin['gender'] == 'M') ? $i18n->t('male') : $i18n->t('female'); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('birthday'); ?></td>
                    <td><?php echo htmlspecialchars($admin['birth_day']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('phone'); ?></td>
                    <td><?php echo htmlspecialchars($admin['phone']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('email'); ?></td>
                    <td><?php echo htmlspecialchars($admin['email']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('admin_level_id'); ?></td>
                    <td><?php echo htmlspecialchars($admin['admin_level_name']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('language_id'); ?></td>
                    <td><?php echo htmlspecialchars($admin['language_id']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('last_reset_password'); ?></td>
                    <td><?php echo htmlspecialchars($admin['last_reset_password']); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('blocked'); ?></td>
                    <td><?php echo $admin['blocked'] ? $i18n->t('yes') : $i18n->t('no'); ?></td>
                </tr>
                <tr>
                    <td><?php echo $i18n->t('active'); ?></td>
                    <td><?php echo $admin['active'] ? $i18n->t('yes') : $i18n->t('no'); ?></td>
                </tr>
                <tr>
                    <td></td>
                    <td>
                        <button type="button" class="btn btn-primary" onclick="window.location='#user-profile-update'"><?php echo $i18n->t('edit'); ?></button>
                        <button type="button" class="btn btn-warning" onclick="window.location='#update-password'"><?php echo $i18n->t('update_password'); ?></button>
                    </td>
                </tr>
            </table>
        </form>
    </div>
<?php
}
}
catch(Exception $e)
{
    echo $e->getMessage();
    exit;
}

File: inc.graphql-resources/backend/go/controller/user_profile.go
package controller

import (
	"context"
	"database/sql"
	"encoding/json"
	"graphqlapplication/constant"
	"graphqlapplication/util"
	"html/template"
	"net/http"
	"path/filepath"
	"strings"
	"time"

	"github.com/gorilla/sessions"
)

// UserProfileHandler handles user profile-related logic.
type UserProfileHandler struct {
	DB    *sql.DB
	Store *sessions.CookieStore
}

// NewUserProfileHandler creates a new instance of UserProfileHandler.
func NewUserProfileHandler(db *sql.DB, store *sessions.CookieStore) *UserProfileHandler {
	return &UserProfileHandler{
		DB:    db,
		Store: store,
	}
}

// AdminProfileData defines the structure for the admin profile data to be displayed.
type AdminProfileData struct {
	AdminID           string
	Name              sql.NullString
	Username          sql.NullString
	Gender            sql.NullString
	BirthDay          sql.NullString // Using string for DATE type for easy formatting
	Email             sql.NullString
	Phone             sql.NullString
	LanguageID        sql.NullString
	LastResetPassword sql.NullString
	Blocked           sql.NullBool
	Active            sql.NullBool
	AdminLevelName    sql.NullString
}

// PageData is the structure for data passed to the HTML template.
type PageData struct {
	Profile      AdminProfileData
	IsUpdateMode bool
	I18n         func(string, ...interface{}) string
	Lang         string
}

// GetProfile handles GET and POST requests to /user-profile.
func (h *UserProfileHandler) GetProfile(w http.ResponseWriter, r *http.Request) {
	session, err := h.Store.Get(r, constant.SessionKey)
	if err != nil {
		http.Error(w, "Failed to get session", http.StatusInternalServerError)
		return
	}

	username, ok := session.Values[constant.SessionUsername].(string)
	if !ok || username == "" {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	if r.Method == http.MethodGet {
		h.handleGetUserProfile(w, r, username)
	} else if r.Method == http.MethodPost {
		h.handlePostUserProfile(w, r, username)
	} else {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleGetUserProfile handles GET requests to display the profile or edit form.
func (h *UserProfileHandler) handleGetUserProfile(w http.ResponseWriter, r *http.Request, username string) {
	// Using the injected h.DB
	ctx := r.Context()

	// Get language from header or query param, default to 'en'
	lang := r.Header.Get("X-Language-Id")
	if lang == "" {
		lang = "en" // Default language
	}
	// Add language to the context so it can be used by util.T
	ctx = context.WithValue(ctx, constant.LanguageKey, lang)

	// Create an i18n closure function for the template
	i18nFunc := func(key string, args ...interface{}) string {
		return util.T(ctx, key, args...)
	}

	query := `
		SELECT 
			a.admin_id, a.name, a.username, a.gender, a.birth_day, a.email, a.phone, 
			a.language_id, a.last_reset_password, a.blocked, a.active,
			al.name AS admin_level_name
		FROM admin a
		LEFT JOIN admin_level al ON a.admin_level_id = al.admin_level_id
		WHERE a.username = ?`

	var profile AdminProfileData
	err := h.DB.QueryRow(query, username).Scan(
		&profile.AdminID, &profile.Name, &profile.Username, &profile.Gender, &profile.BirthDay,
		&profile.Email, &profile.Phone, &profile.LanguageID, &profile.LastResetPassword,
		&profile.Blocked, &profile.Active, &profile.AdminLevelName,
	)

	if err != nil {
		if err == sql.ErrNoRows {
			http.Error(w, i18nFunc("admin_not_found"), http.StatusNotFound)
		} else {
			http.Error(w, i18nFunc("failed_to_fetch_details"), http.StatusInternalServerError)
		}
		return
	}

	isUpdateMode := r.URL.Query().Get("action") == "update"

	// Path to the template
	tmplPath := filepath.Join("template", "user-profile.html")
	tmpl, err := template.New(filepath.Base(tmplPath)).Funcs(template.FuncMap{
		"T":       i18nFunc,
		"ToLower": strings.ToLower,
	}).ParseFiles(tmplPath)

	if err != nil {
		http.Error(w, "Internal Server Error: "+err.Error(), http.StatusInternalServerError)
		return
	}

	pageData := PageData{
		Profile:      profile,
		IsUpdateMode: isUpdateMode,
		I18n:         i18nFunc,
		Lang:         lang,
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	err = tmpl.Execute(w, pageData)
	if err != nil {
		http.Error(w, "Internal Server Error: "+err.Error(), http.StatusInternalServerError)
	}
}

// handlePostUserProfile handles POST requests to update the profile.
func (h *UserProfileHandler) handlePostUserProfile(w http.ResponseWriter, r *http.Request, username string) {
	w.Header().Set("Content-Type", "application/json")

	// Limit body size for security
	r.Body = http.MaxBytesReader(w, r.Body, 10<<20) // 10 MB
	if err := r.ParseMultipartForm(10 << 20); err != nil {
		http.Error(w, "Failed to parse form", http.StatusBadRequest)
		return
	}

	// Using the injected h.DB
	query := `
		UPDATE admin SET 
			name = ?, email = ?, gender = ?, birth_day = ?, phone = ? 
		WHERE username = ?`

	stmt, err := h.DB.Prepare(query)
	if err != nil {
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": "Failed to update profile: " + err.Error()})
		return
	}
	defer stmt.Close()

	// Get values from the form
	name := r.FormValue("name")
	email := r.FormValue("email")
	gender := r.FormValue("gender")
	birthDay := r.FormValue("birth_day")
	phone := r.FormValue("phone")

	_, err = stmt.Exec(name, email, gender, birthDay, phone, username)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": "Failed to update profile: " + err.Error()})
		return
	}

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]interface{}{"success": true, "message": "Profile updated successfully"})
}

// UpdatePassword handles GET and POST requests to /update-password.
func (h *UserProfileHandler) UpdatePassword(w http.ResponseWriter, r *http.Request) {
	session, err := h.Store.Get(r, constant.SessionKey)
	if err != nil {
		http.Error(w, "Failed to get session", http.StatusInternalServerError)
		return
	}

	username, ok := session.Values[constant.SessionUsername].(string)
	if !ok || username == "" {
		http.Error(w, "Unauthorized", http.StatusUnauthorized)
		return
	}

	if r.Method == http.MethodGet {
		h.handleGetUpdatePassword(w, r)
	} else if r.Method == http.MethodPost {
		h.handlePostUpdatePassword(w, r, username)
	} else {
		http.Error(w, "Method not allowed", http.StatusMethodNotAllowed)
	}
}

// handleGetUpdatePassword handles GET requests to display the update password form.
func (h *UserProfileHandler) handleGetUpdatePassword(w http.ResponseWriter, r *http.Request) {
	ctx := r.Context()
	lang := r.Header.Get("X-Language-Id")
	if lang == "" {
		lang = "en"
	}
	ctx = context.WithValue(ctx, constant.LanguageKey, lang)

	i18nFunc := func(key string, args ...interface{}) string {
		return util.T(ctx, key, args...)
	}

	tmplPath := filepath.Join("template", "update-password.html")
	tmpl, err := template.New(filepath.Base(tmplPath)).Funcs(template.FuncMap{
		"T": i18nFunc,
	}).ParseFiles(tmplPath)

	if err != nil {
		http.Error(w, "Internal Server Error: "+err.Error(), http.StatusInternalServerError)
		return
	}

	w.Header().Set("Content-Type", "text/html; charset=utf-8")
	err = tmpl.Execute(w, nil) // No data needed for this template
	if err != nil {
		http.Error(w, "Internal Server Error: "+err.Error(), http.StatusInternalServerError)
	}
}

// handlePostUpdatePassword handles POST requests to update the password.
func (h *UserProfileHandler) handlePostUpdatePassword(w http.ResponseWriter, r *http.Request, username string) {
	w.Header().Set("Content-Type", "application/json")
	ctx := r.Context()
	lang := r.Header.Get("X-Language-Id")
	if lang == "" {
		lang = "en"
	}
	ctx = context.WithValue(ctx, constant.LanguageKey, lang)

	r.ParseForm()
	currentPassword := r.FormValue("current_password")
	newPassword := r.FormValue("new_password")
	confirmPassword := r.FormValue("confirm_password")

	if newPassword == "" {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": util.T(ctx, "new_password_required")})
		return
	}

	if newPassword != confirmPassword {
		w.WriteHeader(http.StatusBadRequest)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": util.T(ctx, "password_mismatch")})
		return
	}

	// Fetch the current user's hashed password
	var dbPassword string
	err := h.DB.QueryRow("SELECT password FROM admin WHERE username = ?", username).Scan(&dbPassword)
	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": util.T(ctx, "failed_to_fetch_details")})
		return
	}

	// Verify the current password
	if util.DoubleSha1(currentPassword) != dbPassword {
		w.WriteHeader(http.StatusUnauthorized)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": util.T(ctx, "incorrect_current_password")})
		return
	}

	// Hash the new password
	newPasswordHash := util.DoubleSha1(newPassword)

	// Update the database
	_, err = h.DB.Exec(
		"UPDATE admin SET password = ?, last_reset_password = ? WHERE username = ?",
		newPasswordHash,
		time.Now().Format(constant.DateTimeFormat),
		username,
	)

	if err != nil {
		w.WriteHeader(http.StatusInternalServerError)
		json.NewEncoder(w).Encode(map[string]interface{}{"success": false, "message": util.T(ctx, "failed_to_update_password")})
		return
	}

	// Invalidate the old session password hash.
	// This is good practice but would require re-login. Let's just return success for now.

	w.WriteHeader(http.StatusOK)
	json.NewEncoder(w).Encode(map[string]interface{}{"success": true, "message": util.T(ctx, "password_updated_successfully")})
}
